File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/stopDetails/StopDetailsFilteredRouteView.kt
package com.mbta.tid.mbta_app.android.stopDetails

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.HorizontalDivider
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.compose.ui.zIndex
import com.mbta.tid.mbta_app.android.ModalRoutes
import com.mbta.tid.mbta_app.android.R
import com.mbta.tid.mbta_app.android.component.HeadsignRowView
import com.mbta.tid.mbta_app.android.component.LineHeader
import com.mbta.tid.mbta_app.android.component.PillDecoration
import com.mbta.tid.mbta_app.android.component.RouteHeader
import com.mbta.tid.mbta_app.android.util.fromHex
import com.mbta.tid.mbta_app.model.Alert
import com.mbta.tid.mbta_app.model.RealtimePatterns
import com.mbta.tid.mbta_app.model.StopDetailsFilter
import com.mbta.tid.mbta_app.model.TripDetailsFilter
import com.mbta.tid.mbta_app.model.response.GlobalResponse
import kotlinx.datetime.Instant

@Composable
fun StopDetailsFilteredRouteView(
    stopId: String,
    viewModel: StopDetailsViewModel,
    global: GlobalResponse?,
    now: Instant,
    stopFilter: StopDetailsFilter,
    tripFilter: TripDetailsFilter?,
    updateStopFilter: (StopDetailsFilter?) -> Unit,
    updateTripFilter: (TripDetailsFilter?) -> Unit,
    openAlertDetails: (ModalRoutes.AlertDetails) -> Unit
) {  val departures = viewModel.stopDepartures.collectAsState().value

    if (departures != null) {
        val patternsByStop = departures.routes.find { it.routeIdentifier == stopFilter.routeId }
        val expectedDirection = stopFilter.directionId
        if (patternsByStop == null) {
            return
        }

        val data =
            departures.stopDetailsFormattedTrips(stopFilter.routeId, stopFilter.directionId, now)

        val alerts: List<Alert> =
            if (global != null) {
                patternsByStop.alertsHereFor(directionId = expectedDirection, global = global)
            } else {
                emptyList()
            }

        val routeHex: String =
            patternsByStop.line?.color ?: patternsByStop.representativeRoute.color
        val routeColor: Color = Color.fromHex(routeHex)
        Box(Modifier.fillMaxSize().background(routeColor)) {
            HorizontalDivider(
                Modifier.fillMaxWidth().zIndex(1f).border(2.dp, colorResource(R.color.halo))
            )
            Column(
                Modifier.fillMaxSize()
                    .padding(horizontal = 16.dp)
                    .verticalScroll(rememberScrollState())
            ) {
                if (patternsByStop.line != null) {
                    LineHeader(patternsByStop.line!!, patternsByStop.routes)
                } else {
                    RouteHeader(patternsByStop.representativeRoute)
                }
                DirectionPicker(patternsByStop, stopFilter, updateStopFilter)

                Spacer(modifier = Modifier.padding(8.dp))

                Column(
                    Modifier.background(colorResource(R.color.fill3), RoundedCornerShape(8.dp))
                ) {
                    for ((index, alert) in alerts.withIndex()) {
                        Column {
                            StopDetailsAlertHeader(
                                alert,
                                routeColor,
                                modifier =
                                Modifier.clickable(
                                    onClickLabel = stringResource(R.string.displays_more_info)
                                ) {
                                    openAlertDetails(
                                        ModalRoutes.AlertDetails(
                                            alertId = alert.id,
                                            lineId = patternsByStop.line?.id,
                                            routeIds = patternsByStop.routes.map { it.id },
                                            stopId = patternsByStop.stop.id
                                        )
                                    )
                                }
                            )
                            if (index < alerts.size - 1 || data.isNotEmpty()) {
                                HorizontalDivider(Modifier.background(colorResource(R.color.halo)))
                            }
                        }
                    }
                    for ((index, row) in data.withIndex()) {
                        val modifier =
                            if (index == 0 || index == data.size - 1)
                                Modifier.background(
                                    colorResource(R.color.fill3),
                                    RoundedCornerShape(8.dp)
                                )
                            else Modifier.background(colorResource(R.color.fill3))

                        val route =
                            patternsByStop.routes.first { it.id == row.upcoming.trip.routeId }

                        Column(modifier.border(1.dp, colorResource(R.color.halo))) {
                            HeadsignRowView(
                                row.upcoming.trip.headsign,
                                RealtimePatterns.Format.Some(listOf(row.formatted), null),
                                Modifier.padding(horizontal = 16.dp, vertical = 10.dp)
                                    .clickable(
                                        onClickLabel = null,
                                        onClick = {
                                            updateTripFilter(
                                                TripDetailsFilter(
                                                    row.upcoming.trip.id,
                                                    row.upcoming.vehicle?.id,
                                                    row.upcoming.stopSequence
                                                )
                                            )
                                        }
                                    ),
                                pillDecoration =
                                if (patternsByStop.line != null)
                                    PillDecoration.OnRow(route = route)
                                else null
                            )
                        }
                    }
                }

                TripDetailsView(
                    tripFilter = tripFilter,
                    stopId = stopId,
                    stopDetailsVM = viewModel,
                    now = now
                )
            }
        }
    }
}

File: iosApp/iosApp/Pages/LegacyStopDetails/StopDetailsFilteredRouteView.swift
//
//  StopDetailsFilteredRouteView.swift
//  iosApp
//
//  Created by Horn, Melody on 2024-04-05.
//  Copyright © 2024 MBTA. All rights reserved.
//

import os
import shared
import SwiftUI

struct StopDetailsFilteredRouteView: View {
    var analytics: StopDetailsAnalytics = AnalyticsProvider.shared
    let patternsByStop: PatternsByStop?
    let alerts: [shared.Alert]
    let downstreamAlerts: [shared.Alert]
    let now: Instant
    var filter: StopDetailsFilter?
    var setFilter: (StopDetailsFilter?) -> Void
    let pushNavEntry: (SheetNavigationStackEntry) -> Void
    let pinned: Bool

    struct RowData {
        let route: Route
        let headsign: String
        let formatted: RealtimePatterns.Format
        let navigationTarget: SheetNavigationStackEntry?

        init(
            route: Route,
            headsign: String,
            formatted: RealtimePatterns.Format,
            navigationTarget: SheetNavigationStackEntry? = nil
        ) {
            self.route = route
            self.headsign = headsign
            self.formatted = formatted
            self.navigationTarget = navigationTarget
        }

        init?(upcoming: UpcomingTrip, route: Route, stopId: String, now: Instant) {
            let formatted = if let formattedUpcomingTrip = RealtimePatterns.companion.formatUpcomingTrip(
                now: now,
                upcomingTrip: upcoming,
                routeType: route.type,
                context: .stopDetailsFiltered
            ) {
                RealtimePatterns.FormatSome(trips: [formattedUpcomingTrip], secondaryAlert: nil)
            } else {
                RealtimePatterns.FormatNone(secondaryAlert: nil)
            }

            if !(formatted is RealtimePatterns.FormatSome) {
                return nil
            }

            self.route = route
            headsign = upcoming.trip.headsign
            self.formatted = formatted

            if let vehicleId = upcoming.prediction?.vehicleId, let stopSequence = upcoming.stopSequence {
                navigationTarget = .tripDetails(
                    tripId: upcoming.trip.id,
                    vehicleId: vehicleId,
                    target: .init(stopId: stopId, stopSequence: stopSequence.intValue),
                    routeId: upcoming.trip.routeId,
                    directionId: upcoming.trip.directionId
                )
            } else {
                navigationTarget = nil
            }
        }
    }

    let rows: [RowData]

    init(
        departures: StopDetailsDepartures,
        global: GlobalResponse?,
        now: Instant,
        filter: StopDetailsFilter?,
        setFilter: @escaping (StopDetailsFilter?) -> Void,
        pushNavEntry: @escaping (SheetNavigationStackEntry) -> Void,
        pinned: Bool
    ) {
        self.filter = filter
        self.setFilter = setFilter
        let patternsByStop = departures.routes.first(where: { $0.routeIdentifier == filter?.routeId })
        self.patternsByStop = patternsByStop
        self.now = now
        self.pushNavEntry = pushNavEntry
        self.pinned = pinned
        let expectedDirection: Int32? = filter?.directionId

        if let patternsByStop, let expectedDirection {
            if let global {
                alerts = patternsByStop.alertsHereFor(directionId: expectedDirection, global: global)
                downstreamAlerts = patternsByStop.alertsDownstream(directionId: expectedDirection)
            } else {
                alerts = []
                downstreamAlerts = []
            }

            var rows: [RowData] = departures.stopDetailsFormattedTrips(
                routeId: patternsByStop.routeIdentifier,
                directionId: expectedDirection,
                filterAtTime: now
            ).compactMap { tripAndFormat in
                let upcoming = tripAndFormat.upcoming
                guard let route = (patternsByStop.routes.first { $0.id == upcoming.trip.routeId }) else {
                    Logger().error("""
                    Failed to find route ID \(upcoming.trip.routeId) from upcoming \
                    trip in patternsByStop.routes (\(patternsByStop.routes.map(\.id)))
                    """)
                    return nil
                }
                return RowData(
                    upcoming: upcoming,
                    route: route,
                    stopId: patternsByStop.stop.id,
                    now: now
                )
            }
            let realtimePatterns = patternsByStop.patterns.filter { $0.directionId() == expectedDirection }
            let statusRows = Self.getStatusDepartures(realtimePatterns: realtimePatterns, now: now)
            rows.append(contentsOf: statusRows)
            self.rows = rows
        } else {
            alerts = []
            downstreamAlerts = []
            rows = []
        }
    }

    static func getStatusDepartures(realtimePatterns: [RealtimePatterns], now: Instant) -> [RowData] {
        StopDetailsDepartures.companion.getStatusDepartures(realtimePatterns: realtimePatterns, now: now)
            .map {
                RowData(
                    route: $0.route,
                    headsign: $0.headsign,
                    formatted: $0.formatted
                )
            }
    }

    var body: some View {
        if let patternsByStop {
            let routeHex: String? = patternsByStop.line?.color ?? patternsByStop.representativeRoute.color
            let routeColor: Color? = if let routeHex { Color(hex: routeHex) } else { nil }
            ZStack {
                if let routeColor {
                    routeColor
                }
                ScrollView {
                    VStack {
                        if let line = patternsByStop.line {
                            LineHeader(line: line, routes: patternsByStop.routes)
                        } else {
                            RouteHeader(route: patternsByStop.representativeRoute)
                        }
                        DirectionPicker(
                            patternsByStop: patternsByStop,
                            filter: filter,
                            setFilter: setFilter
                        ).fixedSize(horizontal: false, vertical: true)

                        ZStack {
                            Color.fill3.ignoresSafeArea(.all)
                            VStack(spacing: 0) {
                                ForEach(Array(alerts.enumerated()), id: \.offset) { index, alert in
                                    VStack(spacing: 0) {
                                        StopDetailsAlertHeader(alert: alert, routeColor: routeColor)
                                            .onTapGesture {
                                                pushNavEntry(.alertDetails(
                                                    alertId: alert.id,
                                                    line: patternsByStop.line,
                                                    routes: patternsByStop.routes
                                                ))
                                                analytics.tappedAlertDetails(
                                                    routeId: patternsByStop.routeIdentifier,
                                                    stopId: patternsByStop.stop.id,
                                                    alertId: alert.id
                                                )
                                            }
                                        if index < alerts.count - 1 || !rows.isEmpty {
                                            Divider().background(Color.halo)
                                        }
                                    }
                                }
                                ForEach(Array(downstreamAlerts.enumerated()), id: \.offset) { index, alert in
                                    VStack(spacing: 0) {
                                        StopDetailsDownstreamAlert(alert: alert, routeColor: routeColor)
                                            .onTapGesture {
                                                pushNavEntry(.alertDetails(
                                                    alertId: alert.id,
                                                    line: patternsByStop.line,
                                                    routes: patternsByStop.routes
                                                ))
                                                analytics.tappedAlertDetails(
                                                    routeId: patternsByStop.routeIdentifier,
                                                    stopId: patternsByStop.stop.id,
                                                    alertId: alert.id
                                                )
                                            }
                                        if index < alerts.count - 1 || !rows.isEmpty {
                                            Divider().background(Color.halo)
                                        }
                                    }
                                }
                                ForEach(Array(rows.enumerated()), id: \.offset) { index, row in
                                    VStack(spacing: 0) {
                                        OptionalNavigationLink(value: row.navigationTarget, action: { entry in
                                            pushNavEntry(entry)
                                            analytics.tappedDepartureRow(
                                                routeId: patternsByStop.routeIdentifier,
                                                stopId: patternsByStop.stop.id,
                                                pinned: pinned,
                                                alert: alerts.count > 0
                                            )
                                        }) {
                                            HeadsignRowView(
                                                headsign: row.headsign,
                                                predictions: row.formatted,
                                                pillDecoration: patternsByStop.line != nil ?
                                                    .onRow(route: row.route) : .none
                                            )
                                        }
                                        .accessibilityInputLabels([row.headsign])
                                        .padding(.vertical, 10)
                                        .padding(.horizontal, 16)

                                        if index < rows.count - 1 {
                                            Divider().background(Color.halo)
                                        }
                                    }
                                }
                            }
                        }
                        .withRoundedBorder()
                    }
                    .padding([.top, .horizontal], 8)
                    .padding([.bottom], 48)
                }
            }.ignoresSafeArea(.all)
        } else {
            EmptyView()
        }
    }
}
